libevmasm/ConstantOptimiser.cpp

```diff
@@ -155,36 +155,72 @@ AssemblyItems CodeCopyMethod::execute(Assembly& _assembly) const
 	bytes data = toBigEndian(m_value);
 	assertThrow(data.size() == 32, OptimizerException, "Invalid number encoding.");
 	AssemblyItems actualCopyRoutine = copyRoutine();
-	actualCopyRoutine[4] = _assembly.newData(data);
+	if (m_params.evmVersion.hasPush0())
+		actualCopyRoutine[3] = _assembly.newData(data);
+	else
+		actualCopyRoutine[4] = _assembly.newData(data);
 	return actualCopyRoutine;
 }
 
-AssemblyItems const& CodeCopyMethod::copyRoutine()
+AssemblyItems const& CodeCopyMethod::copyRoutine() const
 {
-	AssemblyItems static copyRoutine{
-		// constant to be reused 3+ times
-		u256(0),
+	// PUSH0 is cheaper than PUSHn/DUP/SWAP.
+	if (m_params.evmVersion.hasPush0())
+	{
+		// This costs ~29 gas.
+		AssemblyItems static copyRoutine{
+			// back up memory
+			// mload(0)
+			Instruction::PUSH0,
```

Comment:
Could use `u256(0)` in these places, but the explicit instruction felt better for readability and ensuring the `actualCopyRoutine[3]` position is fixed.

Reply:
Fair enough, changed it.

Reply:
After I changed to `u256(0)` a few test cases fail because codecopy does not seem to be selected. That suggest the cost is more than expected, i.e. `u256(0)` generates larger code than `push0`. Need to investigate.

Reply:
@ekpyron I won't have time to debug this for the next 2 weeks. Feel free to take it over.

Reply:
If I understood this right, codecopy cost is more than expected and thus is not being selected because when computing the byte size, `PUSH0` is not taken in account.
At `AssemblyItem::bytesRequired`, the return value for any `PUSHN` is at minimum `2` but for `PUSH0` should be always `1`:
https://github.com/ethereum/solidity/blob/374a6fd50e1bc404ed156e77cc07a63803977622/libevmasm/AssemblyItem.cpp#L126-L127